Source/Protobuf/ArtifactExtensions.cs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Dolittle.SDK.Artifacts;
+using PbArtifact = Dolittle.Artifacts.Contracts.Artifact;
+
+namespace Dolittle.SDK.Protobuf
+{
+    /// <summary>
+    /// Conversion extensions for converting between <see cref="IArtifact"/> and <see cref="PbArtifact"/>.
+    /// </summary>
+    public static class ArtifactExtensions
+    {
+        /// <summary>
+        /// Convert a <see cref="IArtifact"/> to a <see cref="PbArtifact"/>.
+        /// </summary>
+        /// <param name="artifact"><see cref="IArtifact"/> to convert.</param>
+        /// <returns>The converted <see cref="PbArtifact"/>.</returns>
+        public static PbArtifact ToProtobuf(this IArtifact artifact)
+            => new PbArtifact { Id = artifact.Id.ToProtobuf(), Generation = artifact.Generation.Value };
+
+        /// <summary>
+        /// Convert a <see cref="PbArtifact"/> to a <typeparamref name="TArtifact"/>.
+        /// </summary>
+        /// <remarks>
+        /// The method assumes that the <see cref="IArtifact" /> class has a constructor that takes in
+        /// <see cref="ArtifactId" /> as first parameter and <see cref="Generation" /> as second parameter.
+        /// It also assumes that the <see cref="ArtifactId" /> type in the <see cref="IArtifact" /> only has the default parameterless constructor.
+        /// </remarks>
+        /// <param name="artifact"><see cref="PbArtifact"/> to convert.</param>
+        /// <typeparam name="TArtifact">The <see cref="Type" /> of the <see cref="IArtifact" />.</typeparam>
+        /// <returns>The converted <see cref="Artifact{TId}"/>.</returns>
+        public static TArtifact To<TArtifact>(this PbArtifact artifact)
+            where TArtifact : class, IArtifact
+        {
+            var idType = typeof(TArtifact).GetProperty(nameof(IArtifact.Id)).PropertyType;
+            var generationType = typeof(TArtifact).GetProperty(nameof(IArtifact.Generation)).PropertyType;
+            object artifactId;
+            try
+            {
+                artifactId = Activator.CreateInstance(idType);
+                var artifactIdValueProperty = idType.GetProperty(nameof(ArtifactId.Value));
+                artifactIdValueProperty.SetValue(artifactId, artifact.Id.ToGuid());
+            }
+            catch (MissingMethodException)
+            {
+                throw new ArtifactIdDoesNotHaveDefaultParameterlessConstructor(typeof(TArtifact), artifact, idType);
+            }
+
+            Generation generation = artifact.Generation;
+            var constructor = typeof(TArtifact).GetConstructor(new[] { idType, generationType });
+            if (constructor == default) throw new ArtifactDoesNotHaveConstructorWithIdAndGeneration(typeof(TArtifact), artifact);
```

Comment:
```suggestion
                throw new TypeDoesNotHaveDefaultParameterlessConstructor(typeof(TArtifact), artifact, idType);
            }

            Generation generation = artifact.Generation;
            var constructor = typeof(TArtifact).GetConstructor(new[] { idType, generationType });
            if (constructor == default) throw new TypeDoesNotHaveConstructorWithIdAndGeneration(typeof(TArtifact), artifact);
```

As this method is generalized I suggest we should drop the word `Artifact` from the exceptions, eg. calling `pbArtifact.To<EventType>()` and getting a `ArtifactDoesNotHaveConstructorWithIdAndGeneration` sounds a bit weird.
Not sure if the word `Type` is the correct replacement here, it's maybe a bit too vague.